### PR TITLE
[ASTImporter] Fix infinite recurse on return type declared inside body (#68775)

### LINKED ISSUE #68775: [clang][ASTImporter] infinite recursion on lambda with deduced return type from struct declared inside
`ASTNodeImporter::VisitFunctionDecl` considered about infinite recursion on function with deduced return type from struct declared inside function body, but it's not general enough on the following code:
```c
void foo() {
   (void) []() {
     struct X {};
     return X();
   };
}
```
The return type of this lambda is not declared as `auto` so infinite recursion still occurs whem imported:
```c
TEST_P(ASTImporterOptionSpecificTestBase, LambdaReturnWithStructInside) {
  Decl *From, *To;
  std::tie(From, To) = getImportedDecl(
      R"(
      void foo() {
        (void) []() {
          struct X {};
          return X();
        };
      }
      )",
      Lang_CXX11, "", Lang_CXX11, "foo");
  auto *ToLambda = FirstDeclMatcher<LambdaExpr>().match(To, lambdaExpr());
  EXPECT_TRUE(ToLambda);
}
```

### PR DESCRIPTION
**Pull request by danix800**
Lambda without trailing auto could have return type declared inside the body too.

Fixes #68775

### CODE REVIEW

File: clang/lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -3649,19 +3649,15 @@ class IsTypeDeclaredInsideVisitor
 
 /// This function checks if the function has 'auto' return type that contains
 /// a reference (in any way) to a declaration inside the same function.
```
Comment (mizvekov) [inline clang/lib/AST/ASTImporter.cpp:3651]:
```suggestion
/// This function checks if the given function has a return type that contains
/// a reference (in any way) to a declaration inside the same function.
```

**Review by mizvekov — COMMENTED** [review]
I think it would be a good idea to double check this for performance regressions, since this case will recurse into the function every time with this patch.
Though I don't know if there is a better way to test it than to merge it and wait for it to be picked up by google.

Another idea to avoid any potential regressions, is to mark the return type as deduced anyway, with the distinction that 'auto' was not written.

File: clang/lib/AST/ASTImporter.cpp (hunk 1 above)
Comment (danix800) [reply clang/lib/AST/ASTImporter.cpp]: Thanks for reminding, fixed in https://github.com/llvm/llvm-project/pull/89096/commits/3f1d714a2aa4b236afe9bb1384be073fb155a2b8

**Comment by danix800** [comment]
> I think it would be a good idea to double check this for performance regressions, since this case will recurse into the function every time with this patch. Though I don't know if there is a better way to test it than to merge it and wait for it to be picked up by google.

For large TU it's worth considering for performance. The previous commit relaxed the `TypeVisitor` checking to all `FunctionDecl`. I reverted back to the original `auto` impl, plus an extra condition that if it's a `lambda` definition, see https://github.com/llvm/llvm-project/pull/89096/commits/3f1d714a2aa4b236afe9bb1384be073fb155a2b8

> 
> Another idea to avoid any potential regressions, is to mark the return type as deduced anyway, with the distinction that 'auto' was not written.

I tested this idea, might be like this:
```
diff --git a/clang/lib/AST/ASTImporter.cpp b/clang/lib/AST/ASTImporter.cpp
index 712958594473..988efcbef3a3 100644
--- a/clang/lib/AST/ASTImporter.cpp
+++ b/clang/lib/AST/ASTImporter.cpp
@@ -3680,24 +3680,10 @@ ExpectedDecl ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
     // E.g.: auto foo() { struct X{}; return X(); }
     // To avoid an infinite recursion when importing, create the FunctionDecl
     // with a simplified return type.
-    if (hasAutoReturnTypeDeclaredInside(D)) {
-      FromReturnTy = Importer.getFromContext().VoidTy;
-      UsedDifferentProtoType = true;
-    }
-    FunctionProtoType::ExtProtoInfo FromEPI = FromFPT->getExtProtoInfo();
-    // FunctionProtoType::ExtProtoInfo's ExceptionSpecDecl can point to the
-    // FunctionDecl that we are importing the FunctionProtoType for.
-    // To avoid an infinite recursion when importing, create the FunctionDecl
-    // with a simplified function type.
-    if (FromEPI.ExceptionSpec.SourceDecl ||
-        FromEPI.ExceptionSpec.SourceTemplate ||
-        FromEPI.ExceptionSpec.NoexceptExpr) {
-      FunctionProtoType::ExtProtoInfo DefaultEPI;
-      FromEPI = DefaultEPI;
-      UsedDifferentProtoType = true;
-    }
+    FunctionProtoType::ExtProtoInfo DefaultEPI;
+    UsedDifferentProtoType = true;
     FromTy = Importer.getFromContext().getFunctionType(
-        FromReturnTy, FromFPT->getParamTypes(), FromEPI);
+        Importer.getFromContext().VoidTy, FromFPT->getParamTypes(), DefaultEPI);
     FromTSI = Importer.getFromContext().getTrivialTypeSourceInfo(
         FromTy, D->getBeginLoc());
   }
```
but this will break two testcases from `check-clang-astmerge-exprs-cpp`:
```
Failed Tests (2):
  Clang :: ASTMerge/class-template/test.cpp
  Clang :: ASTMerge/exprs-cpp/test.cpp
```

File: clang/lib/AST/ASTImporter.cpp [hunk 2]
```diff
@@ -3647,15 +3647,19 @@ class IsTypeDeclaredInsideVisitor
...
-bool ASTNodeImporter::hasAutoReturnTypeDeclaredInside(FunctionDecl *D) {
+bool ASTNodeImporter::hasReturnTypeDeclaredInside(FunctionDecl *D) {
   QualType FromTy = D->getType();
   const auto *FromFPT = FromTy->getAs<FunctionProtoType>();
   assert(FromFPT && "Must be called on FunctionProtoType");
 
+  bool IsLambdaDefinition = false;
+  if (const auto *MD = dyn_cast<CXXMethodDecl>(D))
+    IsLambdaDefinition = cast<CXXRecordDecl>(MD->getDeclContext())->isLambda();
```
Comment (mizvekov) [inline clang/lib/AST/ASTImporter.cpp:3659]: Thanks, this works, but if we are going this way, why not also check the lambda was defined without a return type?

**Comment by mizvekov** [comment]
> I tested this idea, might be like this:

I meant something else. This patch makes it so the delayed function type importation happens unconditionally.
I was aware that change would cause breakages.

What I meant is that, we have a special sugar type node that marks deduction, this AutoType, and we apply it everywhere else, except this lambda case you exposed. I was going to suggest we do the same for these lambdas. That would make things more consistent.

File: clang/lib/AST/ASTImporter.cpp (hunk 2 above)
Comment (danix800) [reply clang/lib/AST/ASTImporter.cpp:3659]: > Thanks, this works, but if we are going this way, why not also check the lambda was defined without a return type?

Could you please give some direction on how to test **the lambda was defined without a return type**?

Does this mean that we need to check for possible `return` statements within lambda?

Comment (danix800) [reply clang/lib/AST/ASTImporter.cpp:3659]: Or does this mean to test `!FromFPT->hasTrailingReturn()`?

Comment (mizvekov) [reply clang/lib/AST/ASTImporter.cpp:3659]: No, you don't need to hunt for return statements.

You can just check on the CXXMethodDecl that it's FunctionProtoType has a trailing return (`hasTrailingReturn` method).
I think that should cover it for all cases.

Comment (mizvekov) [reply clang/lib/AST/ASTImporter.cpp:3659]: Also, this problem of the return type not being wrapped on an AutoType only happens on lambdas in *C++11*.

Comment (danix800) [reply clang/lib/AST/ASTImporter.cpp:3659]: Got it! Thanks for these useful info!

**Review by mizvekov — APPROVED** [review]
LGTM, Thanks!

**Pull request merged**